src/shared/mount-util.c

```diff
@@ -742,3 +745,228 @@ int mount_option_mangle(
 
         return 0;
 }
+
+int bind_mount_in_namespace(
+                pid_t target,
+                const char *propagate_path,
+                const char *incoming_path,
+                const char *src,
+                const char *dest,
+                int read_only,
+                int make_file_or_directory) {
+
+        _cleanup_close_pair_ int errno_pipe_fd[2] = { -1, -1 };
+        char mount_slave[] = "/tmp/propagate.XXXXXX", *mount_tmp, *mount_outside, *p;
+        bool mount_slave_created = false, mount_slave_mounted = false,
+                mount_tmp_created = false, mount_tmp_mounted = false,
+                mount_outside_created = false, mount_outside_mounted = false;
+        _cleanup_free_ char *chased_src = NULL;
+        struct stat st;
+        pid_t child;
+        int r;
+
+        assert(target > 0);
+        assert(propagate_path);
+        assert(incoming_path);
+        assert(src);
+        assert(dest);
+
+        /* One day, when bind mounting /proc/self/fd/n works across
+         * namespace boundaries we should rework this logic to make
+         * use of it... */
+
+        p = strjoina(propagate_path, "/");
+        if (laccess(p, F_OK) < 0)
+                return log_debug_errno(SYNTHETIC_ERRNO(r == -ENOENT ? -EOPNOTSUPP : r), "Target does not allow propagation of mount points");
```

Comment:
laccess() returns its error in errno, but you check `r` here, which isn't set at all.

That said, I just posted #18257 inspired by this.

Should probably become this
```c
r = laccess(p, F_OK); 
if (r < 0)
        return log_debug_errno(r == -ENOENT ? SYNTHETIC_ERRNO(EOPNOTSUPP) : r, …
```

That is, once #18257 is merged. You could also do it like this:

```
if (laccess(p, F_OK) < 0)
        return log_debug_errno(errno == ENOENT ? SYNTHETIC_ERRNO(EOPNOSTUPP) : errno, …
```

in case you don't want to wait for that PR

Reply:
I'll do the latter for now, thanks

Reply:
The other PR is merged, so rebased to use the new style.